File: gateway/internal/protocol/protocol.go
// Package protocol memuat kontrak pesan WebSocket Humi.
// Sinkron dengan: skill protocol-contract, ai-service/app/protocol.py,
// dan frontend/src/lib/protocol.ts. Ubah SEMUA dalam satu commit.
package protocol

import (
	"encoding/binary"
	"fmt"
)

// Tipe pesan client -> server.
const (
	TypeText       = "text"
	TypeAudioStart = "audio_start"
	TypeAudioChunk = "audio_chunk"
	TypeAudioEnd   = "audio_end"
	TypeInterrupt  = "interrupt"
)

// Tipe pesan server -> client.
const (
	TypeSessionReady = "session_ready"
	TypeSttFinal     = "stt_final"
	TypeLlmSentence  = "llm_sentence"
	TypeTtsStart     = "tts_start"
	TypeTtsEnd       = "tts_end"
	TypeTurnEnd      = "turn_end"
	TypeError        = "error"
)

// Emosi Fish S2 Basic 24 + netral (tanpa tag).
const (
	EmotionNetral     = "netral"
	EmotionHappy      = "happy"
	EmotionSad        = "sad"
	EmotionAngry      = "angry"
	EmotionExcited    = "excited"
	EmotionCalm       = "calm"
	EmotionNervous    = "nervous"
	EmotionConfident  = "confident"
	EmotionSurprised  = "surprised"
	EmotionSatisfied  = "satisfied"
	EmotionDelighted  = "delighted"
	EmotionScared     = "scared"
	EmotionWorried    = "worried"
	EmotionUpset      = "upset"
	EmotionFrustrated = "frustrated"
	EmotionDepressed  = "depressed"
	EmotionEmpathetic = "empathetic"
	EmotionEmbarrassed = "embarrassed"
	EmotionDisgusted  = "disgusted"
	EmotionMoved      = "moved"
	EmotionProud      = "proud"
	EmotionRelaxed    = "relaxed"
	EmotionGrateful   = "grateful"
	EmotionCurious    = "curious"
	EmotionSarcastic  = "sarcastic"
)

// Header audio biner: [seq u32 LE][len u32 LE][payload].
const HeaderBytes = 8

// PackAudio membungkus payload PCM menjadi frame biner.
func PackAudio(seq uint32, payload []byte) []byte {
	frame := make([]byte, HeaderBytes+len(payload))
	binary.LittleEndian.PutUint32(frame[0:4], seq)
	binary.LittleEndian.PutUint32(frame[4:8], uint32(len(payload)))
	copy(frame[HeaderBytes:], payload)
	return frame
}

// UnpackAudio mengurai frame biner. Validasi truncated header/payload.
func UnpackAudio(frame []byte) (uint32, []byte, error) {
	if len(frame) < HeaderBytes {
		return 0, nil, fmt.Errorf("truncated header: got %d bytes, need %d", len(frame), HeaderBytes)
	}
	seq := binary.LittleEndian.Uint32(frame[0:4])
	length := binary.LittleEndian.Uint32(frame[4:8])
	if len(frame) < HeaderBytes+int(length) {
		return 0, nil, fmt.Errorf("truncated payload: header claims %d bytes, frame has %d", length, len(frame)-HeaderBytes)
	}
	payload := frame[HeaderBytes : HeaderBytes+int(length)]
	return seq, payload, nil
}

// IsValidEmotion cek apakah string adalah emosi valid.
func IsValidEmotion(e string) bool {
	switch e {
	case EmotionNetral, EmotionHappy, EmotionSad, EmotionAngry, EmotionExcited, EmotionCalm, EmotionNervous, EmotionConfident, EmotionSurprised, EmotionSatisfied, EmotionDelighted, EmotionScared, EmotionWorried, EmotionUpset, EmotionFrustrated, EmotionDepressed, EmotionEmpathetic, EmotionEmbarrassed, EmotionDisgusted, EmotionMoved, EmotionProud, EmotionRelaxed, EmotionGrateful, EmotionCurious, EmotionSarcastic:
		return true
	default:
		return false
	}
}

File: ai-service/app/protocol.py
from __future__ import annotations

import struct
from typing import Literal

from pydantic import BaseModel

EMOTIONS = (
    "netral",
    "happy",
    "sad",
    "angry",
    "excited",
    "calm",
    "nervous",
    "confident",
    "surprised",
    "satisfied",
    "delighted",
    "scared",
    "worried",
    "upset",
    "frustrated",
    "depressed",
    "empathetic",
    "embarrassed",
    "disgusted",
    "moved",
    "proud",
    "relaxed",
    "grateful",
    "curious",
    "sarcastic",
)

# Framing audio biner: [seq u32 LE][len u32 LE][payload]
_HEADER = struct.Struct("<II")
HEADER_BYTES = _HEADER.size


class Inbound(BaseModel):
    type: Literal["text", "audio_start", "audio_chunk", "audio_end", "interrupt"]
    text: str | None = None
    format: str | None = None
    sample_rate: int | None = None
    seq: int | None = None
    lang: str | None = None


class Config(BaseModel):
    model: str
    voice: str


class Outbound(BaseModel):
    type: Literal[
        "session_ready",
        "stt_final",
        "llm_sentence",
        "tts_start",
        "tts_end",
        "turn_end",
        "error",
    ]
    config: Config | None = None
    text: str | None = None
    emotion: str | None = None
    seq: int | None = None
    format: str | None = None
    sample_rate: int | None = None
    message: str | None = None


def pack_audio(seq: int, payload: bytes) -> bytes:
    return _HEADER.pack(seq, len(payload)) + payload


def unpack_audio(frame: bytes) -> tuple[int, bytes]:
    if len(frame) < HEADER_BYTES:
        raise ValueError(f"truncated header: got {len(frame)} bytes, need {HEADER_BYTES}")
    seq, length = _HEADER.unpack_from(frame)
    if len(frame) < HEADER_BYTES + length:
        raise ValueError(  # noqa: E501
            f"truncated payload: header claims {length} bytes, frame has {len(frame) - HEADER_BYTES}"  # noqa: E501
        )
    payload = frame[HEADER_BYTES : HEADER_BYTES + length]
    return seq, payload
